addons/web/static/src/core/dropdown/dropdown.js

```diff
@@ -1,411 +1,211 @@
-        }
+                if (this.hasParent) {
+                    target.classList.add("o-dropdown--has-parent");
+                }
 
-        useEffect(
-            (isOpen) => {
-                if (isOpen) {
-                    this.props.onOpened();
+                if (this.props.state) {
+                    return;
                 }
-            },
-            () => [this.state.open]
-        );
-    }
 
-    // -------------------------------------------------------------------------
-    // Private
-    // -------------------------------------------------------------------------
+                const clickHandler = () => this.handleClick();
+                const mouseEnterHandler = () => this.handleMouseEnter();
 
-    /**
-     * Changes the dropdown state and notifies over the Dropdown bus.
-     *
-     * All state changes must trigger on the bus, except when reacting to
-     * another dropdown state change.
-     *
-     * @see onDropdownStateChanged()
-     *
-     * @param {Partial<DropdownState>} stateSlice
-     */
-    async changeStateAndNotify(stateSlice) {
-        if (stateSlice.open && this.props.beforeOpen) {
-            await this.props.beforeOpen();
-            if (status(this) === "destroyed") {
-                return;
-            }
-        }
-        // Update the state
-        Object.assign(this.state, stateSlice);
-        // Notify over the bus
-        /** @type DropdownStateChangedPayload */
-        const stateChangedPayload = {
-            emitter: this,
-            newState: { ...this.state },
-        };
-        Dropdown.bus.trigger("state-changed", stateChangedPayload);
-        this.props.onStateChanged({ ...this.state });
-    }
+                target.addEventListener("click", clickHandler);
+                target.addEventListener("mouseenter", mouseEnterHandler);
 
-    /**
-     * Closes the dropdown.
-     *
-     * @returns {Promise<void>}
-     */
-    close() {
-        return this.changeStateAndNotify({ open: false, groupIsOpen: false });
+                return () => {
+                    target.removeEventListener("click", clickHandler);
+                    target.removeEventListener("mouseenter", mouseEnterHandler);
+                };
+            },
+            () => [this.target]
+        );
     }
 
-    /**
-     * Opens the dropdown.
-     *
-     * @returns {Promise<void>}
-     */
-    open() {
-        return this.changeStateAndNotify({ open: true, groupIsOpen: this.props.autoOpen });
+    get hasParent() {
+        return this.nesting.hasParent;
     }
 
-    /**
-     * Toggles the dropdown open state.
-     *
-     * @returns {Promise<void>}
-     */
-    toggle() {
-        const toggled = !this.state.open;
-        return this.changeStateAndNotify({
-            open: toggled,
-            groupIsOpen: toggled && this.props.autoOpen,
-        });
+    get target() {
+        return this.__owl__.bdom?.bdom.el.previousElementSibling;
```

Comment:
Not sure this is a good idea to guess that a button is used as a toggler and set just before the element.
Maybe the solution using slots was already fine to use

-> I guess that you wanted to dissociate the concept of dropdown from the button, but maybe we could create a DropdownButton component, using a button and a the Dropdown? So we better differenciate them depending the use case

Reply:
For me, it adds complexity, because a developer has to know that the Dropdown component can find its toggler by itself. And it's a bit of magic 🧙‍♂️